### PR TITLE
[shelly] Fix NPE on discovery and add RSSI - channel

### PR DESCRIPTION
**Pull request by falkena**
Hello everybody,

this PR fixes NPE i got during discovery after upgrade to openHAB 2.5.2. Additionally, I've added
a channel to read "raw" rssi values provided by shelly devices.

Signed-off-by: Alexander Falkenstern <alexander.falkenstern@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java [hunk 1]
```diff
@@ -471,15 +471,21 @@ private String request(String uri) throws IOException {
...
+                if (message.contains("Timeout") || message.contains("Connection reset")
+                        || type.toLowerCase().contains("timeout")) {
+                    logger.debug("{}: Shelly API timeout ({}), retry", thingName, type);
+                    timeoutErrors++;
+                    retry = true;
+                } else {
+                    throw new IOException(thingName + ": Shelly API call failed (" + type + "), uri=" + uri);
+                }
+            } else {
+                logger.debug("{}: Shelly API can not determinate the exception cause, retry", thingName);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]:
```suggestion
                logger.debug("{}: Shelly API can not determine the exception cause, retry", thingName);
```

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java [hunk 2]
```diff
@@ -471,15 +471,21 @@ private String request(String uri) throws IOException {
         try {
             result = innerRequest(uri);
         } catch (IOException e) {
+            String message = e.getMessage();
             String type = StringUtils.substringAfterLast(e.getCause().toString(), ".");
-            if (e.getMessage().contains("Timeout") || type.toLowerCase().contains("timeout")
-                    || e.getMessage().contains("Connection reset")) {
-                logger.debug("{}: Shelly API timeout ({}), retry", thingName, type);
+            if ((message != null) && (type != null)) {
```
Comment (9037568) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: Extraneous parens

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java (hunk 1 above)
Comment (9037568) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: Also "cannot".

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java [hunk 3]
```diff
@@ -489,14 +495,17 @@ private String request(String uri) throws IOException {
                 timeoutsRecovered++;
                 logger.debug("Shelly API timeout recovered");
             } catch (IOException e) {
+                String message = e.getMessage();
                 String type = StringUtils.substringAfterLast(e.getCause().toString(), ".");
-                if (e.getMessage().contains("Timeout") || type.toLowerCase().contains("timeout")
-                        || e.getMessage().contains("Connection reset")) {
-                    throw new IOException(
-                            thingName + ": Shelly API timeout (" + type + "), uri=" + uri);
+                if ((message != null) && (type != null)) {
```
Comment (9037568) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: Extraneous parens.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java [hunk 4]
```diff
@@ -489,14 +495,17 @@ private String request(String uri) throws IOException {
                 timeoutsRecovered++;
                 logger.debug("Shelly API timeout recovered");
             } catch (IOException e) {
+                String message = e.getMessage();
```
Comment (9037568) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: It appears this logic is used in multiple places. Please create a method for it.

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java (hunk 1 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: determine says the same

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: Thanks for bringing this up. I changed a occurrences where e.getMessage() is used to check on null and map this to ""

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: same, you change would also change the logic

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java (hunk 4 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java]: yep, tried to fixed them all (except some logger.xxx. which is able to handle null)

**Comment by markus7017** [comment]
Thanks for reporting this. I included some changes.

However, some of you changes would change the logic. A channel for the native RSSI value has been removed during initial approval to conform to OH standards. The implemented channel uses the corresponding system type and maps the value to 1-4 depending on the signal strength.

**Comment by falkena** [comment]
Thanks for your feedback. I can't understand, why the channel conflicts with OH standards. But Ok: i'll increase my system load with rest API - calls already performed. For the NPE I#ve raised a bug: https://github.com/openhab/openhab-addons/issues/7069

**Pull request closed without merging**